drivers/rn2xx3/rn2xx3_internal.c

```diff
@@ -0,0 +1,355 @@
+{
+    rn2xx3_t *dev = (rn2xx3_t *)arg;
+    mutex_unlock(&(dev->resp_lock));
+}
+
+static bool _wait_reply(rn2xx3_t *dev, uint8_t timeout)
+{
+    dev->resp_done = 0;
+    dev->resp_size = 0;
+    dev->resp_buf[0] = 0;
+
+    xtimer_ticks64_t sent_time = xtimer_now64();
+
+    xtimer_t resp_timer;
+    resp_timer.callback = isr_resp_timeout;
+    resp_timer.arg = dev;
+
+    xtimer_set(&resp_timer, (uint32_t)timeout * US_PER_SEC);
+
+    /* wait for results */
+    while ((!dev->resp_done) &&
+           xtimer_less(xtimer_diff32_64(xtimer_now64(), sent_time),
+                                        xtimer_ticks_from_usec(timeout * US_PER_SEC))) {
+        mutex_lock(&(dev->resp_lock));
+    }
+
+    xtimer_remove(&resp_timer);
+
+    if (dev->resp_done == 0) {
+        DEBUG("[rn2xx3] response timeout\n");
+        return true;
+    }
+
+    return false;
+}
+
+void rn2xx3_hex_to_bytes(const char *hex, uint8_t *byte_array)
+{
+    const uint8_t charmap[] = {
+        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, /* 01234567 */
+        0x08, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, /* 89:;<=>? */
+        0x00, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x00, /* @ABCDEFG */
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, /* HIJKLMNO */
+    };
+
+    size_t len = strlen(hex);
+    for (uint8_t pos = 0; pos < len; pos += 2) {
+        uint8_t idx0 = ((uint8_t)hex[pos + 0] & 0x1F) ^ 0x10;
+        uint8_t idx1 = ((uint8_t)hex[pos + 1] & 0x1F) ^ 0x10;
+        byte_array[pos / 2] = (uint8_t)(charmap[idx0] << 4) | charmap[idx1];
+    };
+}
+
+void rn2xx3_bytes_to_hex(const uint8_t *byte_array, char *hex, uint8_t max_len)
+{
+    char *pos = &hex[0];
+    const char *hex_chars = "0123456789ABCDEF";
+
+    for (unsigned i = 0; i < max_len; ++i) {
+        *pos++ = hex_chars[(*byte_array >> 4) & 0xF];
+        *pos++ = hex_chars[(*byte_array++) & 0xF];
+    }
+    *pos = '\0';
+}
+
+void rn2xx3_set_internal_state(rn2xx3_t *dev, uint8_t state)
+{
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP || dev->int_state == state) {
+        return;
+    }
+
+#if ENABLE_DEBUG
+    printf("[rn2xx3] new state: ");
+    switch(state) {
+        case RN2XX3_INT_STATE_CMD:
+            printf("CMD");
+            break;
+
+        case RN2XX3_INT_STATE_IDLE:
+            printf("IDLE");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_JOIN:
+            printf("JOIN");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_RX_MESSAGE:
+            printf("RX MSG");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_RX_PORT:
+            printf("RX PORT");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_TX:
+            printf("TX");
+            break;
+
+        case RN2XX3_INT_STATE_RESET:
+            printf("RESET");
+            break;
+
+        case RN2XX3_INT_STATE_SLEEP:
+            printf("SLEEP");
+            break;
+
+        default:
+            printf("UNKNOWN");
+            break;
+    }
+    puts("");
```

Comment:
why not use `puts` in each case instead?

Reply:
indeed that's better